content_scripts/marks.coffee

```diff
@@ -1,45 +1,75 @@
-        scrollX: window.scrollX,
-        scrollY: window.scrollY
-      HUD.showForDuration "Created local mark '#{keyChar}'", 1000
-
-    @remove()
-
-    false
-
-root.activateGotoMode = ->
-  handlerStack.push keydown: (e) ->
-    keyChar = KeyboardUtils.getKeyChar(event)
-    return unless keyChar isnt ""
-
-    if /[A-Z]/.test keyChar
-      chrome.runtime.sendMessage
-        handler: 'gotoMark'
-        markName: keyChar
-    else if /[a-z]/.test keyChar
-      [baseLocation, sep, hash] = window.location.href.split '#'
-      markString = localStorage["vimiumMark|#{baseLocation}|#{keyChar}"]
-      if markString?
-        mark = JSON.parse markString
-        window.scrollTo mark.scrollX, mark.scrollY
-        HUD.showForDuration "Jumped to local mark '#{keyChar}'", 1000
-
-    @remove()
-
-    false
+
+Marks =
+  mode: null
+  previousPosition: null
+
+  exit: (continuation = null) ->
+    @mode?.exit()
+    @mode = null
+    continuation?()
+
+  # This returns the key which is used for storing mark locations in localStorage.
+  getLocationKey: (keyChar) ->
+    "vimiumMark|#{window.location.href.split('#')[0]}|#{keyChar}"
+
+  getMarkString: ->
+    JSON.stringify scrollX: window.scrollX, scrollY: window.scrollY
+
+  setPreviousPosition: ->
+    @previousPosition = @getMarkString()
+
+  showMessage: (message, keyChar) ->
+    HUD.showForDuration "#{message} \"#{keyChar}\".", 1000
+
+  activateCreateMode: ->
+    @mode = new Mode
+      name: "create-mark"
+      indicator: "Create mark..."
+      suppressAllKeyboardEvents: true
+      keypress: (event) =>
+        keyChar = String.fromCharCode event.charCode
+        # If <Shift> is depressed, then it's a global mark, otherwise it's a local mark.  This is consistent
+        # vim's [A-Z] for global marks, [a-z] for local marks.  However, it also admits other non-Latin
+        # characters.
+        @exit =>
+          if event.shiftKey
+              chrome.runtime.sendMessage
+                handler: 'createMark'
+                markName: keyChar
+                scrollX: window.scrollX
+                scrollY: window.scrollY
+              , => @showMessage "Created global mark", keyChar
+          else
+              localStorage[@getLocationKey keyChar] = @getMarkString()
+              @showMessage "Created local mark", keyChar
+
+  activateGotoMode: (registryEntry) ->
+    # We pick off the last character of the key sequence used to launch this command. Usually this is just "`".
+    # We then use that character, so together usually the sequence "``", to jump back to the previous
+    # position.  The "previous position" is recorded below, and is registered via @setPreviousPosition()
+    # elsewhere for various other jump-like commands.
+    previousPositionKey = registryEntry.key[registryEntry.key.length-1..]
+    @mode = new Mode
+      name: "goto-mark"
+      indicator: "Go to mark..."
+      suppressAllKeyboardEvents: true
+      keypress: (event) =>
```

Comment:
> But what is the advantage of `keydown` over `keypress`?

* For `[a-zA-Z0-9]`, what we calculate from `keydown` is the same as what we get from `keypress` on Latin keyboards.
* For a non-Latin keyboard, we will set the Latin equivalent from the `en_US` keyboard, and show them that we've set the mark under that equivalent.
* For users with multiple non-Latin keyboards, or one or more non-Latin keyboards and `en_US` (potentially across computers, since we now sync marks), the same key in the same position will *always* result in the same mark. With `keypress` the user has to switch to the correct layout first (if it's even on the computer they're using) before they can activate the mark.
* Old non-Latin local marks will no longer work (because they were set in this style originally).

@smblott-github it feels wrong to me to do this just so that single-keyboard non-Latin users can have a prettier message in the HUD. Am I just going to have to live with it?